src/v/ssx/include/ssx/sharded_ptr.h

```diff
@@ -83,12 +84,35 @@ class sharded_ptr {
         return reset(std::make_shared<T>(std::forward<Args>(args)...));
     }
 
+    /// updates the managed object u with an update function
+    /// The update function takes a copy of the underlying object and creates a
+    /// new object using the update function. The update is executed inside the
+    /// guard of the update mutex to ensure that updates are executed
+    /// sequentially.
+    /// Must be called on the home shard and is safe to call concurrently.
+    /// returns an ss::broken_semaphore if stop() has been called.
+    template<typename F>
+    requires std::invocable<F, const T&>
+             && std::same_as<T, std::invoke_result_t<F, const T&>>
+    ss::future<> update(F update_func) {
+        assert_shard();
```

Comment:
nit: assert_shard message is a bit confusing `reset must be called on home shard: ` maybe drop the word reset from there